cmd/thanos/query.go

```diff
@@ -404,10 +413,9 @@ func runQuery(
 		)
 
 		g.Add(func() error {
-			statusProber.SetReady()
+			statusProber.Ready()
```

Comment:
Each component actually has a different life-cycle, as you already know, they set their ready state with a different flow. Some of the components have more complicated logic, e.g  `receive`. 

For the rest, just to set a convention I decided to set ready just before the component ready to serve `gRPC` traffic if they serve `gRPC` because `http` starts to serve slightly faster than `gRPC` and usually there are other steps before `gRPC` server initialized.  Otherwise, I stick to the `http` being served since all of them serve HTTP.

Reply:
Furthermore, questions concerning `oklog/run` package, AFAIK whenever an error happens in a `rungroup` all of the error handlers get called sequentially with registration order, and under normal circumstances, this is a very fast process. However, we recently introduced the graceful shutdown of gRPC and HTTP servers. With those changes they can wait for connections to drain, up to `2m` (which is the default query timeout). Since we always register the HTTP server first, I decided to set the probe to not ready, when HTTP server gets shut down, and stop receiving traffic.

Reply:
Thanks for clearing this up! LGTM :+1:

Reply:
Plus it's worth to note that if HTTP server is drained, we are not ready anyway.